the-collab-lab/tcl-7-smart-shopping-list
# Mm nk join existing list

## Description
This PR adds the functionality that allows a person to join an existing shopping list. It does this by adding a `ShareList` module to handle a person joining a list.

The component handles the following processes:
1. Clicking "Join existing shopping list" should take us to `./join-existing`
2. `./join-existing` we should have the capacity to enter a share code
3. When share is submitted we need the capacity to validate share code
    - if share code is valid we need to:
        -  Set token of list to `localStorage`
        - Route to `./list-view`
    - If share is invalid we pop up an alert window

Additionally the route `.join-existing` was added to our app along with an updated `href` property to the "join an existing shopping list" link found at route `/`. 

### Collection Validation & Routing logic
The component `ShareList` in  `Components/shareList.js`  contains the routing logic within the method `getExistingList` on [line 17](https://github.com/the-collab-lab/tcl-7-smart-shopping-list/blob/d787c3988824a193841f672f4f63070ae06372e4/src/Components/shareList.js#L17). 
1. We create an interface with our Cloud Firestore service and set it to the variable `db`
2. We take the value from the input value that is stored in the components state `this.state.token` pass it as the collection path to the [method `db.collection`](https://firebase.google.com/docs/reference/js/firebase.firestore.Firestore#collection) which gets a [`CollectionReference` instance](https://firebase.google.com/docs/reference/js/firebase.firestore.CollectionReference) based on this path. We then chain a [`.get` method](https://firebase.google.com/docs/reference/js/firebase.firestore.CollectionReference#get) which query's the database and returns a [promise-like object `QuerySnapshot`](https://firebase.google.com/docs/reference/js/firebase.firestore.QuerySnapshot). This is stored to the constant `promise` on [line 19](https://github.com/the-collab-lab/tcl-7-smart-shopping-list/blob/d787c3988824a193841f672f4f63070ae06372e4/src/Components/shareList.js#L19). 😳well that was a lot...
2.  Now we're working asynchronously, we chain a `.then` method which returns a `QuerySnapshot` object which we pass as an argument to the callback function passed to the `.then` method. 
3. In the callback function we house the pass-fail logic triggered by the return value of `!querySnapshot.empty` 
    - If the QuerySnapshot object is empty (no documents) we know it doesn't exist and show an alert.
   - Otherwise we will call `setLocalToken` and pass `this.state.token` in as the token to set in `localStorage`. We then route the client to `./list-view` which uses the `token` stored in `localStorage` to retrieve and serve the list.

The event handler `handleCheckShareCode` is registered to the `onClick` event of the "Join Shopping List" button.

`handleCheckShareCode` prevents the default behavior of the form and performs the runs `getExistingList`.

### Styling of `./join-existing`
We copied the existing style and structure of `SignIn` and added the Google Material style-like inputs that hover on focus. 

Loom demo: https://www.loom.com/share/89c426933b0147ff823377c5321869a4

Hover input:
![image](https://user-images.githubusercontent.com/3664197/82704923-84936380-9c2b-11ea-84e8-ff192280a8dc.png)


### Accessibility
We passed all tests and fulfilled the tabbing requirements.
 ![image](https://user-images.githubusercontent.com/3664197/82701968-f4065480-9c25-11ea-8af6-3fddf0c47a71.png)

## Related Issue
Closes #6 

## Acceptance Criteria

- When the user does not already have a token in localStorage, on the onboarding/home screen, the user can enter an existing token
- Entering the token and hitting submit saves the token to localStorage, effectively giving them joint control of the list
- On submit, show an error if the token does not exist
- If they get an error message, allow them to try again or create a new list


## Type of Changes



|     | Type                       |
| --- | -------------------------- |
|    | :bug: Bug fix              |
|  ✓ | :sparkles: New feature     |
|  ✓  | :hammer: Refactoring       |
|    | :100: Add tests            |
|    | :link: Update dependencies |
|    | :scroll: Docs              |

## Updates

### Before
`./joing-existing` did not exist

### After
Loom demo: https://www.loom.com/share/89c426933b0147ff823377c5321869a4

![image](https://user-images.githubusercontent.com/3664197/82704901-78a7a180-9c2b-11ea-8eab-b57b1278865c.png)






## Testing Steps / QA Criteria



Setup:
1. On your localhost in chrome open the developers tools
2. Select the “Application” tab in the console's top menu.
3. Select “Local Storage” in the console's left menu.
4. Right-click and click clear to delete the local storage. ![image](https://user-images.githubusercontent.com/3664197/82071958-ad47b600-968b-11ea-8ac9-8eca98fdc870.png)
5. In your browser navigate to root `/`
6. you should see ![image](https://user-images.githubusercontent.com/3664197/82070803-1cbca600-968a-11ea-8393-0e6b57178b62.png)

Test 1:
1. On root `/` click `join existing shopping list` 
![image](https://user-images.githubusercontent.com/3664197/82708678-b5779680-9c33-11ea-9d1c-3f1fa2d40e92.png)
2. If navigates to `./join-existing` then passed

Test 2: 
1. On `./join-existing` click `share code` and enter `not a real token` 
![image](https://user-images.githubusercontent.com/3664197/82708750-e061ea80-9c33-11ea-9898-e5a2da138cf1.png)
2. Click `Join shopping list` 
![image](https://user-images.githubusercontent.com/3664197/82708788-f4a5e780-9c33-11ea-8e3b-db82675ad45c.png)
3. If alert pops up then passed 
![image](https://user-images.githubusercontent.com/3664197/82708813-05eef400-9c34-11ea-8cd6-11853dcd7285.png)

Test 3: 
1. On `./join-existing` click `share code` and enter `user`
![image](https://user-images.githubusercontent.com/3664197/82708872-2dde5780-9c34-11ea-805f-8b1f1b9283d6.png)
2. Click `Join shopping list`
3. Passes if:
    - Navigates to `./list-view` 
![image](https://user-images.githubusercontent.com/3664197/82708935-536b6100-9c34-11ea-88ec-deeb6504209a.png)
    - Entering `window.localStorage.getItem('token') === 'user'` into the console returns `true`. 
![image](https://user-images.githubusercontent.com/3664197/82709083-aba26300-9c34-11ea-8b3c-6518b0fdb929.png)
